clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -426,8 +426,10 @@ boolean joinGroupIfNeeded(final Timer timer) {
                     needsJoinPrepare = true;
                 } else {
                     log.info("Generation data was cleared by heartbeat thread. Initiating rejoin.");
-                    resetStateAndRejoin();
-
+                    synchronized (AbstractCoordinator.this) {
```

Comment:
Resetting state should be protected with synchronization.

Reply:
Why don't we make `resetStateAndRejoin` synchronized?

Reply:
In other two callers of `resetStateAndRejoin` it is already synchronized at the caller, so I just follow the same pattern here.

Reply:
Yes, but that is why this was a bug, right?

Reply:
The original bug was on `generation`, but I think we should also be careful on `state` as well.  I thought doing the synchronization on the caller is fine but maybe just enforce it at the callee is better for future-proof, changing it then.